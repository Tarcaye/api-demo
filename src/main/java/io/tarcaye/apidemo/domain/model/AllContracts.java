package io.tarcaye.apidemo.domain.model;

import java.util.Optional;

public interface AllContracts {
    Optional<Contracts> findFor(String customerId);
}
