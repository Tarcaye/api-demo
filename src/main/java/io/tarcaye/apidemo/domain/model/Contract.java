package io.tarcaye.apidemo.domain.model;

public class Contract {

    public enum Type {
        LOCATION, ACHAT
    }

    private final long id;
    private final Customer customer;
    private final Type type;

    public Contract(Customer customer, long id, Type type) {
        this.id = id;
        this.customer = customer;
        this.type = type;
    }

    public long getId() {
        return id;
    }

    public Type getType() {
        return type;
    }

    public boolean belongsTo(Customer customer) {
        return this.customer.getId().equals(customer.getId());
    }
}
