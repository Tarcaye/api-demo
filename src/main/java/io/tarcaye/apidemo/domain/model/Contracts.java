package io.tarcaye.apidemo.domain.model;

import java.util.List;

public class Contracts {

    private final List<Contract> values;

    public Contracts(List<Contract> values) {
        this.values = values;
    }

    public List<Contract> asList() {
        return values;
    }
}
