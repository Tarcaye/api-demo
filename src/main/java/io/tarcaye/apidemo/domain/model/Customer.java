package io.tarcaye.apidemo.domain.model;

public class Customer {
    private final Long id;

    private Customer(Long id) {
        this.id = id;
    }

    public static Customer of(String id) {
        return new Customer(Long.valueOf(id));
    }

    public Long getId() {
        return id;
    }
}
