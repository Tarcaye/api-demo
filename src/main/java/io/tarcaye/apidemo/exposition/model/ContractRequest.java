package io.tarcaye.apidemo.exposition.model;

public class ContractRequest {
}
