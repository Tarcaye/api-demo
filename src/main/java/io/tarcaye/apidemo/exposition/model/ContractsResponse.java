package io.tarcaye.apidemo.exposition.model;

import io.tarcaye.apidemo.domain.model.Contract;
import io.tarcaye.apidemo.domain.model.Contracts;
import lombok.NoArgsConstructor;

import java.util.List;

import static java.util.Collections.emptyList;
import static java.util.Collections.unmodifiableList;
import static java.util.stream.Collectors.toList;

public class ContractsResponse {

    private final List<ContractResponse> contracts;

    public ContractsResponse(List<ContractResponse> contracts) {
        this.contracts = contracts;
    }
    public ContractsResponse() {
        this.contracts = emptyList();
    }

    public static ContractsResponse of(Contracts contracts) {
        List<ContractResponse> contractResponses = contracts.asList().stream()
                .map(ContractResponse::from)
                .collect(toList());

        return new ContractsResponse(contractResponses);
    }

    public List<ContractResponse> getContracts() {
        return unmodifiableList(contracts);
    }

    public static class ContractResponse {
        private final long id;
        private final Type type;

        public ContractResponse() {
            id=0;
            type = null;
        }
        public ContractResponse(long id, Type type) {

            this.id = id;
            this.type = type;
        }

        public static ContractResponse from(Contract contract) {
            return new ContractResponse(113L, Type.ACHAT);
        }

        public long getId() {
            return id;
        }

        public Type getType() {
            return type;
        }

        enum Type {
            ACHAT
        }
    }
}
