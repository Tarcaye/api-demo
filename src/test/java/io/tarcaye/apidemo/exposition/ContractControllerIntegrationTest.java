package io.tarcaye.apidemo.exposition;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.tarcaye.apidemo.ApiDemoApplication;
import io.tarcaye.apidemo.exposition.model.ContractsResponse;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.json.JsonParseException;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.test.context.web.WebAppConfiguration;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;
import org.approvaltests.Approvals;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@AutoConfigureMockMvc
@SpringBootTest(classes = ApiDemoApplication.class)
class ContractControllerIntegrationTest {

    @Autowired
    private MockMvc mockMvc;


    protected String mapToJson(Object obj) throws JsonProcessingException {
        ObjectMapper objectMapper = new ObjectMapper();
        return objectMapper.writeValueAsString(obj);
    }
    protected <T> T mapFromJson(String json, Class<T> clazz)
            throws JsonParseException, JsonMappingException, IOException {

        ObjectMapper objectMapper = new ObjectMapper();
        return objectMapper.readValue(json, clazz);
    }

    @Test
    void getContracts() throws Exception {
        String uri = "/customers/9001/contracts";
        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        int status = mvcResult.getResponse().getStatus();
        assertEquals(200, status);
        String content = mvcResult.getResponse().getContentAsString();
        Approvals.verify(content);
    }

    @Test
    void addContract() throws Exception {
        String uri = "/customers/21/contracts";
        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        int status = mvcResult.getResponse().getStatus();
        assertEquals(200, status);
        String content = mvcResult.getResponse().getContentAsString();
        Approvals.verify(content);
    }




    @Test
    void addAndRetrieveContracts() throws Exception {
        String uri = "/customers/42/contracts";
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        int status = mvcResult.getResponse().getStatus();
        assertEquals(200, status);
        String content = mvcResult.getResponse().getContentAsString();
        Approvals.verify(content);
    }
}