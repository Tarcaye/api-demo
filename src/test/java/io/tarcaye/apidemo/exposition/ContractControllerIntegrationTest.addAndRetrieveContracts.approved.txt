{"contracts":[{"id":113,"type":"ACHAT"},{"id":113,"type":"ACHAT"},{"id":113,"type":"ACHAT"},{"id":113,"type":"ACHAT"}]}